File: colorfilter/paper/Deconvolution/ftvi.m
function x = ftvi(y,k1,k2,k3,mu)

pc = fpc(k1,k2,k3,size(y));

out_it = 8; beta_s = 4;
beta = (mu*100) / (beta_s).^out_it; 

%%%%%
pad = pc{1};
y = padimg(y,pad);

%%% Precompute stuff
[num2 den1 den2] = precomp(y,pc);

% Init
x = y; fct = 0.5;
for i = 1:3
  x(:,:,i) = real(ifft2(fct*num2(:,:,i) ./ (den1 + fct*den2(:,:,i))));
end;

for ot = 1:out_it
  % W sub-problem
  wx = diff(x,1,2)/sqrt(2); wy = diff(x,1,1)/sqrt(2);

  %-- Shrink
  mx = max(10^-4,max(abs(wx(:)))); mx = max(mx,max(abs(wy(:))));
  [thr a b] = gets(beta,mx);
  wx(abs(wx) < thr) = 0; wy(abs(wy) < thr) = 0;
  
  for i = 1:3
    wx(:,:,i) = max(0,a*abs(wx(:,:,i))+b) .* sign(wx(:,:,i));
    wy(:,:,i) = max(0,a*abs(wy(:,:,i))+b) .* sign(wy(:,:,i));
  end;
  
  % X sub-problem
  for i = 1:3
    num1 = [-wx(:,1,i) -diff(wx(:,:,i),1,2) wx(:,end,i)];
    num1 = num1 + [-wy(1,:,i); -diff(wy(:,:,i),1,1); wy(end,:,i)];
    tmp = fft2(num1/sqrt(2)) + mu/beta*num2(:,:,i);
    tmp = tmp ./ (den1 + mu/beta*den2(:,:,i));
    x(:,:,i) = real(ifft2(tmp)); 
  end;
    
  
  beta = beta*beta_s;  % Beta continuation
end;


% Done
x = x(1+pad:end-pad,1+pad:end-pad,:);

function [num2 den1 den2] = precomp(y,pc)

den1 = pc{5}; den2 = pc{6};

k1f = pc{2};
k2f = pc{3};
k3f = pc{4};

num2 = conj(k1f).*fft2(y(:,:,1));
num2(:,:,2) = conj(k2f).*fft2(y(:,:,2));
num2(:,:,3) = conj(k3f).*fft2(y(:,:,3));

function y2 = padimg(y,pad)
y2 = zeros(size(y) + [2*pad 2*pad 0]);
wt = linspace(0,1,2*pad);
wtx = repmat(wt(:)',[size(y,1) 1]); wty = repmat(wt(:),[1 size(y2,2)]);
for i = 1:3
  yi = y(:,:,i);
    
  px = (1-wtx).*repmat(yi(:,end),[1 2*pad])+(wtx).*repmat(yi(:,1),[1 2*pad]);
  yi = [px(:,pad+1:end) yi px(:,1:pad)];
  
  py = (1-wty).*repmat(yi(end,:),[2*pad 1])+(wty).*repmat(yi(1,:),[2*pad 1]);
  yi = [py(pad+1:end,:); yi; py(1:pad,:)];
  
  y2(:,:,i) = yi;
end;


% Compute linear approximation to gradient shrinkage
function [thr a b] = gets(beta,mx)

v = linspace(0,mx,5000);
v2 = shrinkv23(v,beta);
idx = min(find(v2 > 0));

if length(idx) == 0
  a = 0; b = 0; thr = 10;
  return
end;

thr = v(idx);

a = (v2(end)-v2(idx)) / (v(end)-v(idx));
b = v2(end) - a*v(end);

File: colorfilter/paper/Deconvolution/ftvi.py
"""
Explanation:
Function Definitions:
ftvi(y, k1, k2, k3, mu): Main function to perform the optimization.
precomp(y, pc): Precomputes necessary components.
pad_image(y, pad): Pads the image to handle boundary conditions.
gets(beta, mx): Computes thresholds for shrinkage.
shrinkv23(v, beta): Placeholder function for gradient shrinkage. Implement this function as needed.
fpc(k1, k2, k3, y_shape): Placeholder for the fpc function, which needs to be implemented based on the actual function details.
Kernel Operations:
Uses fft2 and ifft2 for Fourier transforms.
np.diff computes gradients similar to MATLAB's diff.
Padding:
Padding is done with linear interpolation to handle edge effects.
Initialization and Iteration:
The optimization is performed through iterative updates, including gradient calculations and kernel adjustments.
Note:

The shrinkv23 and fpc functions are placeholders and should be implemented based on your specific requirements or the MATLAB counterparts.
Make sure to test the Python implementation with your actual data and kernel definitions for accurate results.
"""

import numpy as np
import cv2
from scipy.fft import fft2, ifft2
from scipy.ndimage import convolve

def ftvi(y, k1, k2, k3, mu):
    pc = fpc(k1, k2, k3, y.shape)
    
    out_it = 8
    beta_s = 4
    beta = (mu * 100) / (beta_s ** out_it)

    # Pad image
    pad = pc[0]
    y = pad_image(y, pad)

    # Precompute
    num2, den1, den2 = precomp(y, pc)

    # Initialize
    x = np.copy(y)
    fct = 0.5
    for i in range(3):
        x[:, :, i] = np.real(ifft2(fct * num2[:, :, i] / (den1 + fct * den2[:, :, i])))

    for _ in range(out_it):
        # W sub-problem
        wx = np.diff(x, 1, axis=1) / np.sqrt(2)
        wy = np.diff(x, 1, axis=0) / np.sqrt(2)

        # Shrink
        mx = np.max([1e-4, np.max(np.abs(wx)), np.max(np.abs(wy))])
        thr, a, b = gets(beta, mx)
        wx[np.abs(wx) < thr] = 0
        wy[np.abs(wy) < thr] = 0

        for i in range(3):
            wx[:, :, i] = np.maximum(0, a * np.abs(wx[:, :, i]) + b) * np.sign(wx[:, :, i])
            wy[:, :, i] = np.maximum(0, a * np.abs(wy[:, :, i]) + b) * np.sign(wy[:, :, i])

        # X sub-problem
        for i in range(3):
            num1 = np.pad(np.diff(wx[:, :, i], 1, axis=1), ((0, 0), (0, 1)), mode='constant') - wx[:, :, i]
            num1 += np.pad(np.diff(wy[:, :, i], 1, axis=0), ((0, 1), (0, 0)), mode='constant') - wy[:, :, i]
            tmp = fft2(num1 / np.sqrt(2)) + mu / beta * num2[:, :, i]
            tmp /= (den1 + mu / beta * den2[:, :, i])
            x[:, :, i] = np.real(ifft2(tmp))

        beta *= beta_s  # Beta continuation

    # Remove padding
    x = x[pad:-pad, pad:-pad, :]

    return x

def precomp(y, pc):
    k1f = pc[1]
    k2f = pc[2]
    k3f = pc[3]
    den1 = pc[5]
    den2 = pc[6]

    num2 = np.zeros_like(y)
    num2[:, :, 0] = np.conj(k1f) * fft2(y[:, :, 0])
    num2[:, :, 1] = np.conj(k2f) * fft2(y[:, :, 1])
    num2[:, :, 2] = np.conj(k3f) * fft2(y[:, :, 2])

    return num2, den1, den2

def pad_image(y, pad):
    padded_shape = (y.shape[0] + 2 * pad, y.shape[1] + 2 * pad, y.shape[2])
    y2 = np.zeros(padded_shape)

    for i in range(3):
        yi = y[:, :, i]
        # Pad horizontally
        px = (1 - np.linspace(0, 1, 2 * pad)[:, None]) * np.tile(yi[:, -1:], (1, 2 * pad)) + np.linspace(0, 1, 2 * pad)[:, None] * np.tile(yi[:, :1], (1, 2 * pad))
        yi = np.hstack([px[:, pad:], yi, px[:, :pad]])

        # Pad vertically
        py = (1 - np.linspace(0, 1, 2 * pad)[None, :]) * np.tile(yi[-1:, :], (2 * pad, 1)) + np.linspace(0, 1, 2 * pad)[None, :] * np.tile(yi[:1, :], (2 * pad, 1))
        yi = np.vstack([py[pad:, :], yi, py[:pad, :]])

        y2[:, :, i] = yi

    return y2

def gets(beta, mx):
    v = np.linspace(0, mx, 5000)
    v2 = shrinkv23(v, beta)
    idx = np.min(np.where(v2 > 0)[0]) if len(np.where(v2 > 0)[0]) > 0 else 0

    thr = v[idx] if idx != 0 else 10
    a = (v2[-1] - v2[idx]) / (v[-1] - v[idx]) if idx != 0 else 0
    b = v2[-1] - a * v[-1] if idx != 0 else 0

    return thr, a, b

def shrinkv23(v, beta):
    return np.maximum(v - beta, 0)  # Placeholder; implement the actual shrinkage function

def fpc(k1, k2, k3, y_shape):
    # Placeholder for the fpc function which computes and returns necessary parameters
    # Implement based on your actual 'fpc' function details
    return [0, fft2(k1), fft2(k2), fft2(k3), np.ones(y_shape[:2]), np.ones(y_shape[:2])]

# Example usage
y = np.random.rand(100, 100, 3)  # Example image
k1 = np.ones((5, 5))  # Example kernels
k2 = np.ones((5, 5))
k3 = np.ones((5, 5))
mu = 0.1  # Regularization parameter
x = ftvi(y, k1, k2, k3, mu)

